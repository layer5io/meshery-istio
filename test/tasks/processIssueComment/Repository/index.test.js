const nock = require('nock')

const Repository = require('../../../../src/tasks/processIssueComment/Repository')

const gitGetRefdata = require('../../../fixtures/git.getRef.json')
const gitCreateRefdata = require('../../../fixtures/git.createRef.json')
const reposUpdateFiledata = require('../../../fixtures/repos.updateFile.json')
const pullsCreatedata = require('../../../fixtures/pulls.create.json')
const mockGithub = require('../../../mocks/mockGithub')

describe('Repository', () => {
    const repository = new Repository({
        repo: 'all-contributors-bot',
        owner: 'all-contributors',
        github: mockGithub,
        defaultBranch: 'master',
    })

    const verifyBody = body => {
        expect(body).toMatchSnapshot()
        return true
    }

    test('create new file', async () => {
        nock('https://api.github.com')
            .put(
                `/repos/all-contributors/all-contributors-bot/contents/notes/hello.txt`,
                verifyBody,
            )
            .reply(201, reposUpdateFiledata)

        await repository.createFile({
            content: 'Hello World',
            filePath: 'notes/hello.txt',
            branchName: 'all-contributors/add-jakebolam',
        })
    })

    test('createPullRequest with files', async () => {
        nock('https://api.github.com')
            .get(
                `/repos/all-contributors/all-contributors-bot/git/refs/heads/master`,
            )
            .reply(200, gitGetRefdata)

        nock('https://api.github.com')
            .post(
                `/repos/all-contributors/all-contributors-bot/git/refs`,
                verifyBody,
            )
            .reply(201, gitCreateRefdata)

        nock('https://api.github.com')
            .put(
                `/repos/all-contributors/all-contributors-bot/contents/.all-contributorsrc`,
                verifyBody,
            )
            .reply(200, reposUpdateFiledata)

        nock('https://api.github.com')
            .put(
                `/repos/all-contributors/all-contributors-bot/contents/README.md`,
                verifyBody,
            )
            .reply(200, reposUpdateFiledata)

        nock('https://api.github.com')
            .put(
                `/repos/all-contributors/all-contributors-bot/contents//nested-folder/SOME-DOC.md`,
                verifyBody,
            )
            .reply(200, reposUpdateFiledata)

        nock('https://api.github.com')
            .post(
                `/repos/all-contributors/all-contributors-bot/pulls`,
                verifyBody,
            )
            .reply(201, pullsCreatedata)

        const {
            pullRequestURL,
            pullCreated,
        } = await repository.createPullRequestFromFiles({
            title: 'Pull request title',
            body: 'Pull request body',
            filesByPath: {
                '.all-contributorsrc': {
                    content: JSON.stringify({ test: 'test content' }),
                    originalSha: 'aa218f56b14c9653891f9e74264a383fa43fefbe',
                },
                'README.md': {
                    content: 'Updated list',
                    originalSha: 'aa218f56b14c9653891f9e74264a383fa43fefbl',
                },
                '/nested-folder/SOME-DOC.md': {
                    content: 'Updated list in nested folder',
                    originalSha: 'aa218f56b14de653891f9e74264a383fa43fefbd',
                },
            },
            branchName: 'all-contributors/add-jakebolam',
        })
        expect(pullRequestURL).toEqual(
            'https://github.com/all-contributors/all-contributors-bot/pull/1347',
        )
        expect(pullCreated).toBe(true)
    })
})
