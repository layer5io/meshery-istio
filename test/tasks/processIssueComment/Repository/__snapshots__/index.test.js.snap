// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository create new file 1`] = `
Object {
  "branch": "all-contributors/add-jakebolam",
  "content": "SGVsbG8gV29ybGQ=",
  "message": "docs: create notes/hello.txt [skip ci]",
}
`;

exports[`Repository createPullRequest with files 1`] = `
Object {
  "ref": "refs/heads/all-contributors/add-jakebolam",
  "sha": "aa218f56b14c9653891f9e74264a383fa43fefbd",
}
`;

exports[`Repository createPullRequest with files 2`] = `
Object {
  "branch": "all-contributors/add-jakebolam",
  "content": "eyJ0ZXN0IjoidGVzdCBjb250ZW50In0=",
  "message": "docs: update .all-contributorsrc [skip ci]",
  "sha": "aa218f56b14c9653891f9e74264a383fa43fefbe",
}
`;

exports[`Repository createPullRequest with files 3`] = `
Object {
  "branch": "all-contributors/add-jakebolam",
  "content": "VXBkYXRlZCBsaXN0",
  "message": "docs: update README.md [skip ci]",
  "sha": "aa218f56b14c9653891f9e74264a383fa43fefbl",
}
`;

exports[`Repository createPullRequest with files 4`] = `
Object {
  "branch": "all-contributors/add-jakebolam",
  "content": "VXBkYXRlZCBsaXN0IGluIG5lc3RlZCBmb2xkZXI=",
  "message": "docs: update /nested-folder/SOME-DOC.md [skip ci]",
  "sha": "aa218f56b14de653891f9e74264a383fa43fefbd",
}
`;

exports[`Repository createPullRequest with files 5`] = `
Object {
  "base": "master",
  "body": "Pull request body",
  "head": "all-contributors/add-jakebolam",
  "maintainer_can_modify": true,
  "title": "Pull request title",
}
`;
